### DISCUSSION TITLE
General page cache for different queries

### DISCUSSION THREAD

**Discussion by SahilAggarwal2004** [DuCanhGH/next-pwa#119] [Ideas]
Let's assume that I am building a note taking application and I cached a '/notes' page which has 2 filters 'all' and 'todo'. The 2 urls formed will be '/notes?filter=all' ('/notes' also shows all notes by default) and '/notes?filter=todo' respectively.

Currently the problem with next-pwa or even with this fork is that if '/notes' page is cached only that particular page is shown when offline and if you go to '/notes?filter=todo' and reload the page, it throws the offline fallback page instead to linking it with the cached '/notes' page.

Now here if you have static queries, you can cache '/notes', '/notes?filter=all', '/notes?filter=todo' pages but it will take 3 times the necessary cache. And if you have dynamic queries, you can't even cache the pages.

So I was wondering if it is possible that all the '/notes?query=something' requests can be linked with the cached '/notes' page?

Note that I am taking about a CSR page and thus the page will be same irrespective of the filters and then the data can be fetched, or taken from localStorage (if offline) based on the filter.

**Comment by DuCanhGH** [comment]
Well it depends on whether you use InjectManifest or GenerateSW. But, looking at the question, I'm assuming you are using GenerateSW, in which case, you can do something like this:
```js
const withPWA = require("@ducanh2912/next-pwa").default({
  dest: "public",
  extendDefaultRuntimeCaching: true,
  workboxOptions: {
    runtimeCaching: [
      {
        urlPattern: ({ request, url: { pathname }, sameOrigin }) =>
          request.headers.get("RSC") === "1" &&
          request.headers.get("Next-Router-Prefetch") === "1" &&
          sameOrigin &&
          pathname === "/notes",
        handler: "NetworkFirst",
        options: {
          cacheName: "notes-pages-rsc-prefetch",
          expiration: {
            maxEntries: 32,
            maxAgeSeconds: 24 * 60 * 60,
          },
          matchOptions: {
            ignoreSearch: true,
          },
        },
      },
      {
        urlPattern: ({ request, url: { pathname }, sameOrigin }) =>
          request.headers.get("RSC") === "1" &&
          sameOrigin &&
          pathname === "/notes",
        handler: "NetworkFirst",
        options: {
          cacheName: "notes-pages-rsc",
          expiration: {
            maxEntries: 32,
            maxAgeSeconds: 24 * 60 * 60,
          },
          matchOptions: {
            ignoreSearch: true,
          },
        },
      },
      {
        urlPattern: ({ url: { pathname }, sameOrigin }) =>
          sameOrigin && pathname === "/notes",
        handler: "NetworkFirst",
        options: {
          cacheName: "notes-pages",
          expiration: {
            maxEntries: 32,
            maxAgeSeconds: 24 * 60 * 60,
          },
          matchOptions: {
            ignoreSearch: true,
          },
        },
      },
    ],
  },
});
```
Note that this only matches the page '/notes', so you'd probably want to change `urlPattern`. The idea remains the same, however, which is setting `matchOptions.ignoreSearch` to `true`.

Sorry for not replying, but this is because I'm not notified of the creation of a new discussion, and I've been busy with Serwist (the successor of this fork).

**Reply by SahilAggarwal2004** [reply]
Yeah, I am planning to shift to Serwist now. Thanks for your wonderful libraries!

Continuing the discussion,
Sorry that I didn't mentioned it before, but I am actually precaching the routes with the help of additionalManifestEntries and not at runtime.
And the idea is to avoid multiple caches in case of just query changes

**Reply by SahilAggarwal2004** [reply]
Nevermind, I just saw an option of ignoreURLParametersMatching in the precacheOptions in installSerwist function exported @serwist/sw. I am just confused on how to actually use it since there is no clear documentation available.

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 that will change how `@serwist/precaching` does its things, but it will not prevent `@serwist/precaching` from caching `/notes` as separate pages during runtime. I'd suggest doing this:
```js
import { defaultCache } from "@serwist/next/browser";
import type { PrecacheEntry } from "@serwist/precaching";
import { installSerwist } from "@serwist/sw";

declare const self: ServiceWorkerGlobalScope & {
  // Change this attribute's name to your `injectionPoint`.
  __SW_MANIFEST: (PrecacheEntry | string)[] | undefined;
};

installSerwist({
  precacheEntries: self.__SW_MANIFEST,
  precacheOptions: {
    ignoreURLParametersMatching: [/^filter$/],
  },
  skipWaiting: true,
  clientsClaim: true,
  navigationPreload: true,
  runtimeCaching: [
    {
      urlPattern: ({ request, url: { pathname }, sameOrigin }) =>
        request.headers.get("RSC") === "1" && request.headers.get("Next-Router-Prefetch") === "1" && sameOrigin && pathname === "/notes",
      handler: "NetworkFirst",
      options: {
        cacheName: "notes-pages-rsc-prefetch",
        expiration: {
          maxEntries: 32,
          maxAgeSeconds: 24 * 60 * 60,
        },
        matchOptions: {
          ignoreSearch: true,
        },
      },
    },
    {
      urlPattern: ({ request, url: { pathname }, sameOrigin }) => request.headers.get("RSC") === "1" && sameOrigin && pathname === "/notes",
      handler: "NetworkFirst",
      options: {
        cacheName: "notes-pages-rsc",
        expiration: {
          maxEntries: 32,
          maxAgeSeconds: 24 * 60 * 60,
        },
        matchOptions: {
          ignoreSearch: true,
        },
      },
    },
    {
      urlPattern: ({ url: { pathname }, sameOrigin }) => sameOrigin && pathname === "/notes",
      handler: "NetworkFirst",
      options: {
        cacheName: "notes-pages",
        expiration: {
          maxEntries: 32,
          maxAgeSeconds: 24 * 60 * 60,
        },
        matchOptions: {
          ignoreSearch: true,
        },
      },
    },
    ...defaultCache,
  ],
});
```
Now, if you don't use `runtimeCaching` at all then you can safely remove it from the code above. I'm sorry that there's no documentation available for now, but you can use Workbox's docs as an alternative :)

**Reply by SahilAggarwal2004** [reply]
Thanks a lot for helping out! Although, I figured it out and thought that ignoring every query param will be a good idea.

ignoreURLParametersMatching: [/.*/]

This is what I finally implemented

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 alright :)

**Reply by SahilAggarwal2004** [reply]
Sorry to disturb again, but I had another doubt.

I saw that in serwist, the compiled sw.js is being cached too, where it was not being cached in next-pwa. Is there any way to opt out as I don't really see any advantage of caching it since it's already installed.

Currently I am using a filter as below:
precacheEntries: self.__SW_MANIFEST.filter(({ url }) => !url.includes('/sw.js'))

Is there any better alternative available?

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 use this:
```js
import withSerwistInit from "@serwist/next";

const withSerwist = withSerwistInit({
  cacheOnFrontEndNav: true,
  swSrc: "app/sw.ts",
  swDest: "public/sw.js",
  exclude: [/public\/sw.js/]
});
```
That should do it for you :) This happens because we use `@serwist/webpack-plugin`'s InjectManifest, which creates a child compiler on Next.js's webpack compiler, whereas in `next-pwa`, we use `workbox-webpack-plugin`'s GenerateSW, which creates a Rollup instance instead.

**Reply by SahilAggarwal2004** [reply]
Yeah, that seems cool! Didn't though of that

**Comment by SahilAggarwal2004** [comment]
> @SahilAggarwal2004 use this:
> 
> ```js
> import withSerwistInit from "@serwist/next";
> 
> const withSerwist = withSerwistInit({
>   cacheOnFrontEndNav: true,
>   swSrc: "app/sw.ts",
>   swDest: "public/sw.js",
>   exclude: [/public\/sw.js/]
> });
> ```
> 
> That should do it for you :) This happens because we use `@serwist/webpack-plugin`'s InjectManifest, which creates a child compiler on Next.js's webpack compiler, whereas in `next-pwa`, we use `workbox-webpack-plugin`'s GenerateSW, which creates a Rollup instance instead.

Unfortunately, this is not working :(

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 weird, can you send me the output code? Also set `exclude` to `[/public\/sw.js/, (e) => (console.log(e), false)]` to double check :)

**Reply by SahilAggarwal2004** [reply]
Well that logged out a lot of stuff that I am unable to grab at once😂

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 you gotta try that on a minimal reproduction then 💀

**Reply by SahilAggarwal2004** [reply]
console logs related to sw.js:

`valueCacheVersions` map:
valueCacheVersions: {
  ...,
  'webpack/DefinePlugin self.__SERWIST_SW_ENTRY.sw' => "'/sw.js'",
  ...
}

`assets` object:
assets: {
  '../public/sw.js': [CachedSource],
  ...
}

`assetsInfo` map:
assetsInfo: {
  '../public/sw.js' => [Object],
  ...
}

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 oops, I meant `[/public\/sw.js/, (e) => (console.log(e.asset.name), false)]`. Sorry about that 🥲

**Reply by SahilAggarwal2004** [reply]
exclude: ['../public/sw.js']

This worked out! Assest name is "../public/sw.js"

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 yeah, that should also work :) I suggested using a RegExp lest that path somehow changes in the future, but that's okay.

**Reply by SahilAggarwal2004** [reply]
So stupid of me, I just realised that I was enclosing the regex in double quotes(") 😭

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 
![fucking hell](https://github.com/DuCanhGH/next-pwa/assets/75556609/de199d7f-3996-4b0f-8e98-727b174295d8)

**Reply by SahilAggarwal2004** [reply]
😅
